### ISSUE TITLE
WSL Doesn't run proper binary according to $PATH or `which`

### ISSUE THREAD

**Issue by daryltucker** [microsoft/WSL#8609]
### Version

Microsoft Windows [Version 10.0.22000.795]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.10.16.3-microsoft-standard-WSL2

### Distro Version

Linux WinMachine 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 GNU/Linux

### Other Software

Debian WSL
GNU bash, version 4.4.12(1)-release (x86_64-pc-linux-gnu)

### Repro Steps

Compile Python and install it.
https://gist.github.com/daryltucker/ff8c82c2c1b8092884dd53861cbec4da#file-build-python3-wsl-sh

```bash
✘ 12:16 daryl@WinMachine ~/Projects/cpython $ which python3
/usr/local/bin/python3
 [ v3.10.5 | ✔  ]
✔ 12:16 daryl@WinMachine ~/Projects/cpython $ python3 --version
Python 3.5.3
 [ v3.10.5 | ✔  ]
✔ 12:16 daryl@WinMachine ~/Projects/cpython $ /usr/local/bin/python3 --version
Python 3.10.5
```

### Expected Behavior

I expect `which` to show me which binary is going to be run when I run a command.  Very core and basic functionality for Linux/BASH.

1.  I have no way of knowing which binary is going to run, as I should with `which` on Linux
2. I have no way of knowing which binary is being run instead of the correct binary.
  * Well, there are ways but wanted to expose this as well.

### Actual Behavior

WSL runs some random binary.  No idea which one.  `$PATH` isn't honored properly, and `which` returns erroneous results.

**Comment by daryltucker** [comment]
```
daryl@WinMachine:~$ echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib
```

```
daryl@WinMachine:~$ file /usr/bin/python
/usr/bin/python: symbolic link to /etc/alternatives/python
daryl@WinMachine:~$ file /usr/local/bin/python3
/usr/local/bin/python3: symbolic link to python3.10
```


If you run the following command, it will shift on you, however:
```
daryl@WinMachine:~$ sudo update-alternatives --install /usr/bin/python python /usr/local/bin/python3.10 10

daryl@WinMachine:~$ which python
/usr/bin/python
daryl@WinMachine:~$ python --version
Python 3.10.5
```

To be clear, this is still an issue and `which` should report the correct binary to be run.

**Comment by elsaco** [comment]
At the end of `build-python3-wsl.sh` the new Python build is added with priority 10:

```
sudo update-alternatives --install /usr/bin/python3 python /usr/local/bin/python3.10 10
sudo update-alternatives --install /usr/bin/python python3 /usr/local/bin/python3.10 10
```

If you have another Python binary with a lower priority, it will be used instead. What is the output of `update-alternatives --display python` and `update-alternatives --display python3`?

**Comment by daryltucker** [comment]
Hey, sorry for the confusion.

I reported the bug, and **then** I started playing with `update-alternatives`.  I've removed references to `update-alternatives` from the linked gist/script.

I was asked an about the actual paths of the binaries.  I tried to provide the information, but wanted to preface with a disclaimer that *I had made changes since reporting the issue, so the information could be invalid/different*.

---

I was able to replicate the issue using a fresh WSL install.  I had to use another machine, because `wsl` doesn't support something like `wsl --install Debian -t Debian2`? 

### Attempt to replicate
#### Get a fresh Debian or Ubuntu WSL installation:
```cmd
wsl --install -d Debian
```

#### Update to stable
```bash
sudo apt update; sudo apt upgrade -y
sudo apt install vim
sudo vim /etc/apt/sources.list
```

```bash
cat /etc/apt/sources.list
deb http://deb.debian.org/debian bullseye main
deb http://deb.debian.org/debian bullseye-updates main
deb http://security.debian.org/debian-security bullseye-security main
```

```bash
sudo apt update; sudo apt upgrade -y
sudo apt dist-upgrade
sudo apt install -y python3
```

#### Compile and install OpenSSL and Python:

```bash
sudo apt install -y git build-essential checkinstall pkg-config zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libsqlite3-dev libreadline-dev libffi-dev libbz2-dev libc6-dev tk-dev liblzma-dev uuid-dev curl wget

FILE="/etc/ld.conf.so"
LINE="/usr/local/lib"
grep -qF -- "$LINE" "$FILE" || echo "$LINE" | sudo tee -a "$FILE"

mkdir -p ~/src

cd ~/src
git clone https://github.com/openssl/openssl.git; cd openssl
git fetch -a
LATEST_TAG=$(git tag -l --format="%(creatordate:raw) %(refname)" | grep "OpenSSL_1"| sort -r | head -n1 | cut -d' ' -f3)
git checkout "$LATEST_TAG"
./config shared --prefix=/usr/local/
make -j $(expr $(nproc --all) - 2)
sudo ldconfig

cd ~/src
git clone https://github.com/python/cpython; cd cpython
git fetch -a
LATEST_TAG=$(git tag -l --format="%(creatordate:raw) %(refname)" | grep -v "rc" | grep 'v3.' | grep -E '.*\.[0-9]+$'| sort -r | head -n1 | cut -d' ' -f3)
git checkout "$LATEST_TAG"
LDFLAGS="$(pkg-config --libs openssl)" ./configure --enable-optimizations --with-lto --enable-shared
make -j $(expr $(nproc --all) - 2)
sudo make install
sudo ldconfig
```

#### Whitness
```bash
daryl@WinMachine2:~/src/cpython$ which python3
/usr/local/bin/python3
daryl@WinMachine2:~/src/cpython$ python3 --version
Python 3.9.2
daryl@WinMachine2:~/src/cpython$ /usr/local/bin/python3 --version
Python 3.10.5
```

**Comment by elsaco** [comment]
Look at the output of `which -a python3` and `whereis python3`. Also check `$PATH`. 

There is `python 3.9.2` package you installed after `apt dist-upgrade` that created the link:

```
tux@RIPPER:~$ ls -l /usr/bin/python3
lrwxrwxrwx 1 root root 9 Apr  5  2021 /usr/bin/python3 -> python3.9
```

And you have manually installed `python3.10` in `/usr/local/bin`. 

Did not repro.  Here's output after installing python3.10:

```
tux@RIPPER:~$ which -a python3
/usr/local/bin/python3
/usr/bin/python3
```


```
tux@RIPPER:~$ whereis python3
python3: /usr/bin/python3.9 /usr/bin/python3 /usr/lib/python3.5 /usr/lib/python3.9 /usr/lib/python3 /etc/python3.5 /etc/python3.9 /etc/python3 /usr/local/bin/python3.10-config /usr/local/bin/python3 /usr/local/bin/python3.10 /usr/local/lib/python3.9 /usr/local/lib/python3.10 /usr/share/python3 /usr/share/man/man1/python3.1.gz

```

```
tux@RIPPER:~$ type python3
python3 is hashed (/usr/local/bin/python3)
tux@RIPPER:~$ python3 --version
Python 3.10.
```

**Comment by daryltucker** [comment]
I was able to replicate for a third time, using my instructions step by step.

I am reporting this problem here because I've never experienced this before.  It could be true that this isn't WSL specifically, but I cannot replicate using Debian images installed to VM and going from 9.8 (`stretch`) to Bullseye.


I found this an interesting change:
```
daryl@WinMachine2:~/src/cpython$ command -v python3
/usr/local/bin/python3
daryl@WinMachine2:~/src/cpython$ which python3
/usr/local/bin/python3
daryl@WinMachine2:~/src/cpython$ python3 --version
Python 3.10.5
```

Seems that running `command -v python3` "fixes" the issue.
